### ISSUE TITLE
Allow users to choose between `json` and `jsonb` in Postgres helpers

### ISSUE THREAD

**Issue by sds** [kysely-org/kysely#395]
_Very_ excited for JSON-related helpers added in 7215cc3c4df2a96eedd68d409f63b3350d159c58 to now come built-in to Kysely!

One potential enhancement would be to include `json` variants alongside the `jsonb` forms of the helpers. 

Depending on what you need the result for, it can be significantly more efficient to use `json` (especially if you are returning a larger result set) since Postgres doesn't need to perform extra work to construct the tree.

Once you start JSONifying over 100 rows (or wide result sets with many columns) the performance difference becomes noticeable.

Happy to add this if you are supportive. The one point worth clarifying is if we are willing to change the names of the existing helpers (e.g. `jsonObjectFrom`) to have `jsonb` in the name (e.g. `jsonbObjectFrom`) so that the `json` versions can be easily differentiated.

**Comment by koskimas** [comment]
I wasn't aware of any noticeable speed difference between them. In that case, is there actually ever a good reason to use `jsonb` in those helpers? `jsonb` makes sense as a column type, but as a temporary type, we might just ditch it in favor of `json`? I'm not a big fan of adding both versions.

**Comment by sds** [comment]
**TL;DR**: Both are worth supporting, but if you had to choose, `json` is the better choice for performance reasons and typical use cases.

---

The answer is nuanced, but if the intended purpose is to return the result to a client (and not extract JSON properties for comparison in some other query) then `json` is _usually_ all you need.

If you need to extract values to join or otherwise compare against, then `jsonb` is necessary because any time Postgres actually needs to parse JSON (for example whenever using the [JSON operators](https://www.postgresql.org/docs/15/functions-json.html)) it needs to do the work to construct the `jsonb` tree anyway.

Here's a brief summary if helpful:
 Purpose | `json` | `jsonb` | Comments
---------|-------|---------|-----------
Remove duplicate keys | | ✅ | May be cases where you want to preserve the underlying data regardless of whether or not it is valid, and so `json` may still be preferred.
Prevent invalid Unicode surrogate pairs | | ✅ | See the comments about RFC 7159 [here](https://www.postgresql.org/docs/15/datatype-json.html), but if you're taking data from an external source and dropping it straight into a `json` column, you run the risk of surprising errors when you later attempt to parse the stored value into `jsonb`. This is a reason why it's usually a good idea to use `jsonb` as the column's type, unless you **know** that you're never going to query anything in the JSON document and will always return the JSON string as-is.
Extract property for comparison in query |  | ✅ | If you need to pull out a value for any reason, then you might as well return it as `jsonb` since that's what Postgres will do in order to extract the value, e.g. `WHERE json_column ->> 'property' = '...'`
You just want a JSON string | ✅ | | If none of the above scenarios apply, and you know you just want JSON sent to the client, then `json` is going to be more performant. 

**Note**: if you already know the shape of the JSON you want to output, it's even faster to concatenate strings manually (think [`fast-json-stringify`](https://github.com/fastify/fast-json-stringify)):
```sql
select json_agg(agg) from (select a, b from table limit 10000) as agg;
```
vs:
```sql
select '{"a":' || a || ',"b":"' || b || '"}' from table limit 10000;
```

The latter is faster, though it assumes you don't have any string values with characters that need escaping, e.g. `"`. In those cases, you could wrap string columns with `to_json` if you wanted complete safety, but the more you do this the less performance benefit you'll see.

```sql
select '{"a":' || a || ',"b":"' || to_json(b) || '"}' from table limit 10000;
```

Let me know if I can clarify anything further.

**Comment by koskimas** [comment]
Actually I'll write some benchmarks myself to see what kind of difference we are dealing with here. A is faster than B means nothing if the difference is one millionth of the overall query execution time.

**Comment by koskimas** [comment]
That's 10 milliseconds per 10000 rows. So 1 microsecond per row 😄

**Comment by sds** [comment]
I wasn't trying to emphasize that this only occurs with thousands of rows.

There are many situations where you are selecting more than two columns (which is what the example above was demonstrating for sake of simplicity). Meaningful (milliseconds) differences in execution time can be shown in under 100 rows (i.e. the size of a typical page of a response), but it's highly dependent on your data set.

### `json_agg`
```
explain analyze select json_agg(agg) from (select * from table_a join table_b on table_a.id = table_b.id limit 100) as agg;
...
Execution Time: 2.375 ms
```

### `jsonb_agg`
```
> explain analyze select jsonb_agg(agg) from (select * from table_a join table_b on table_a.id = table_b.id limit 100) as agg;
...
Execution Time: 5.219 ms
```

_Note: `select *` is to keep the example simple—I'm not suggesting anyone do this in their queries._

All else being equal, if performance is relevant, you'd rather use `json` than `jsonb`. Especially if you are implementing the pattern discussed in the [relations recipe](https://github.com/kysely-org/kysely/blob/9325e36613df9ab89ac7333347c886faa1e1cfe8/site/docs/recipes/relations.md) (awesome pattern), where you might have multiple relations all combined into a single result.

**Comment by koskimas** [comment]
My point was more this: With 10k rows (or 100 rows with a huge amount of columns) other parts of the query will still be slow. The total time of a request consists of (at least) these parts:

1. Query execution time (what you are measuring)
2. Time it takes for node to parse the result from the DB (`JSON.parse` + rest of the stuff)
3. Time it takes to serialize the result as the response (`JSON.stringify`)
4. Network delays (although this is a separate issue)

I'd guess `JSON.parse` and `JSON.stringify` are _at least_ as slow as the query. So by saving 2ms on the query, you're actually only making a small part of the request a tiny bit faster. Overall, the users probably won't notice a difference.

Having said all this, if there is absolutely zero reasons to keep `jsonb`, we can switch to `json`. But even the tiniest reason to keep `jsonb` will be more important than this performance difference.

**Comment by koskimas** [comment]
Switched to using `json` instead of `jsonb` in the helpers.

**Comment by virtuallyunknown** [comment]
This may or may not be worth your attention (it's probably a niche case ?), but I've discovered that `.distinct()` will fail here because you can't use `DISTINCT` with `json_build_object`, but you can with `jsonb_build_object`.

```sql
-- ✅ works

SELECT
  json_agg(DISTINCT jsonb_build_object('id', "players"."id")) AS "whatever"
FROM
  "players"

-- ❌ error: could not identify an equality operator for type json

SELECT
  json_agg(DISTINCT json_build_object('id', "players"."id")) AS "whatever"
FROM
  "players"
```

https://kyse.link/NEkuw

A super easy workaround here is to just grab the helper from [the source code](https://github.com/kysely-org/kysely/blob/master/src/helpers/postgres.ts#L158-L168) and modify it to use `jsonb_build_object` (just have to add 1 extra character).

**Comment by emulienfou** [comment]
Would be nice to let us choose between `json` and `jsonb` in helpers like mentioned above.

I have a case with `UNION` and subqueries when if there is no data in DB from a subquery and failed to returns an empty array by example. Got error message like `Error: could not identify an equality operator for type json`

<details>

<summary>Complex query example to reproduce/debug:</summary>

### You can add a header

```sql
select "Expense"."id",
       "Expense"."amount",
       "Expense"."type",
       "Expense"."notes",
       cast("Expense"."datePaid" as text)                                                                     as "date",
       "User"."id"                                                                                            as "ownerId",
       "User"."name"                                                                                          as "owner",
       "User"."image"                                                                                         as "ownerImage",
       'expense'                                                                                              as "_transactionType",
       (select coalesce(json_agg(agg), '[]')
        from (select "File"."id", "File"."url" from "File" where "File"."expenseId" = "Expense"."id") as agg) as "files"
from "Expense"
         left join "Owner" on "Owner"."id" = "Expense"."paidBy"
         left join "User" on "User"."id" = "Owner"."userId"
union
(select "Income"."id",
        "Income"."amount",
        NULL                                                                                                 as "type",
        "Income"."notes",
        cast("Income"."dateReceived" as text)                                                                as "date",
        "User"."id"                                                                                          as "ownerId",
        "User"."name"                                                                                        as "owner",
        "User"."image"                                                                                       as "ownerImage",
        'income'                                                                                             as "_transactionType",
        (select coalesce(json_agg(agg), '[]')
         from (select "File"."id", "File"."url" from "File" where "File"."incomeId" = "Income"."id") as agg) as "files"
 from "Income"
          left join "Owner" on "Owner"."id" = "Income"."receivedBy"
          left join "User" on "User"."id" = "Owner"."userId")
order by "date" desc
limit 25 offset 0;
```

</details>

**Comment by virtuallyunknown** [comment]
My memory is a bit rusty, but AFAIK the real issue is when you need the `eb.fn` json helpers but for json**b**, because they support chaining and they are context/query/type aware.

For example, there is no way (at least to my knowledge) to create `eb.fn.jsonbAgg` and `eb.fn.toJsonb` helpers because you'd have to modify library code itself.

I created a thread about it on the Discord servers and one of the devs told me they *could* add them, so @koskimas if you're still on board with this idea, having those helpers would be really appreciated!

**Comment by p-bakker** [comment]
Couldn't find the thread mentioned above and this case is closed

Is there still some effort or otherwise an open case to get jsonb variants of the helpers?

My usecase: being able to concate 2 jsonb objects using the `||` operator: only works with jsonb, so I cannot use any of the built-in json helpers...
